src/v_fmt.c

```diff
@@ -0,0 +1,544 @@
+
+    spng_set_option(ctx, SPNG_KEEP_UNKNOWN_CHUNKS, 1);
+
+    spng_set_png_buffer(ctx, buffer, buffer_length);
+
+    struct spng_ihdr ihdr = {0};
+    int ret = spng_get_ihdr(ctx, &ihdr);
+
+    if (ret)
+    {
+        I_Printf(VB_ERROR, "V_CachePatchNum: spng_get_ihdr %s\n",
+                 spng_strerror(ret));
+        goto error;
+    }
+
+    if (ihdr.color_type != SPNG_COLOR_TYPE_INDEXED || ihdr.bit_depth != 8)
+    {
+        I_Printf(VB_ERROR, "V_CachePatchNum: Only 8-bit paletted PNG supported");
+        goto error;
+    }
+
+    int color_key = NO_COLOR_KEY;
+    struct spng_trns trns = {0};
+    ret = spng_get_trns(ctx, &trns);
+
+    if (ret && ret != SPNG_ECHUNKAVAIL)
+    {
+        I_Printf(VB_ERROR, "V_CachePatchNum: spng_get_trns %s",
+                 spng_strerror(ret));
+        goto error;
+    }
+
+    for (int i = 0; i < trns.n_type3_entries; ++i)
+    {
+        if (trns.type3_alpha[i] < 255)
+        {
+            color_key = i;
+            break;
+        }
+    }
+
+    int leftoffset = 0, topoffset = 0;
+    uint32_t n_chunks = 0;
+    ret = spng_get_unknown_chunks(ctx, NULL, &n_chunks);
+
+    if (ret && ret != SPNG_ECHUNKAVAIL)
+    {
+        I_Printf(VB_ERROR, "V_CachePatchNum: spng_get_unknown_chunks %s",
+                 spng_strerror(ret));
+        goto error;
+    }
+
+    if (n_chunks > 0)
+    {
+        struct spng_unknown_chunk *chunks = malloc(n_chunks * sizeof(*chunks));
+        spng_get_unknown_chunks(ctx, chunks, &n_chunks);
+        for (int i = 0; i < n_chunks; ++i)
+        {
+            if (!memcmp(chunks[i].type, "grAb", 4) && chunks[i].length == 8)
+            {
+                int *p = chunks[i].data;
+                leftoffset = SWAP_BE32(p[0]);
+                topoffset = SWAP_BE32(p[1]);
+                break;
+            }
+        }
+        free(chunks);
+    }
+
+    size_t image_size = 0;
+    ret = spng_decoded_image_size(ctx, SPNG_FMT_PNG, &image_size);
+
+    if (ret)
+    {
+        I_Printf(VB_ERROR, "V_CachePatchNum: spng_decoded_image_size %s",
+                 spng_strerror(ret));
+        goto error;
+    }
+
+    byte *image = malloc(image_size);
+
+    ret = spng_decode_image(ctx, image, image_size, SPNG_FMT_PNG, 0);
+
+    if (ret)
+    {
+        I_Printf(VB_ERROR, "V_CachePatchNum: spng_decode_image %s",
+                 spng_strerror(ret));
+        free(image);
+        goto error;
+    }
+
+    spng_ctx_free(ctx);
+    Z_Free(buffer);
+
+    patch_t *patch;
+    if (color_key == NO_COLOR_KEY)
+    {
+        patch = V_LinearToPatch(image, ihdr.width, ihdr.height, tag,
+                                &lumpcache[lump]);
+    }
+    else
+    {
+        patch = V_LinearToTransPatch(image, ihdr.width, ihdr.height, color_key,
+                                     tag, &lumpcache[lump]);
+    }
+    patch->leftoffset = leftoffset;
+    patch->topoffset = topoffset;
+    free(image);
+
+    return lumpcache[lump];
+
+error:
```

Comment:
Is it ok to have two jump marks of the same name in the same source file?

Reply:
Just looked it up, `goto` has function scope. Got confused by it anyway. 😉

Reply:
Yes, in C labels are local to a single function. There are `setjmp()`/`longjmp()` for "non-local jumps".